tests/containers/podmansh.pm

```diff
@@ -0,0 +1,200 @@
+
+use Mojo::Base 'containers::basetest';
+use testapi;
+use serial_terminal 'select_serial_terminal';
+use Utils::Systemd qw(systemctl);
+use utils;
+use containers::common;
+
+my $quadlet_dir = '/etc/containers/systemd';
+
+my $src_image = "registry.opensuse.org/opensuse/leap:15.6";
+my $quadlet_container = <<_EOF_;
+[Unit]
+Description=The podmansh container
+After=local-fs.target
+
+[Container]
+Image=$src_image
+ContainerName=podmansh
+UserNS=keep-id
+RunInit=yes
+DropCapability=all
+NoNewPrivileges=true
+
+# change homedir to match 'user-console' needle
+Environment=HOME=/home
+WorkingDir=/home
+
+# OpenQA serial terminal access
+AddDevice=/dev/ttyS0
+
+Exec=sleep infinity
+
+[Quadlet]
+# avoid infinite waiting for network by podman-user-wait-network-online.service
+# https://github.com/systemd/systemd/issues/28762
+DefaultDependencies=false
+
+[Install]
+RequiredBy=default.target
+_EOF_
+
+my $unit_name = 'podmansh.container';
+my $user;
+my $uid;
+my $systemd_user_path;
+my $initial_user_shell;
+
+sub run {
+    my ($self) = @_;
+    select_serial_terminal;
+
+    my $podman = $self->containers_factory('podman');
+
+    # prepare user for rootless containers
+    prepare_rootless($self);
+
+    install_packages('podmansh');
+
+    # check quadlet works
+    my $quadlet = '/usr/libexec/podman/quadlet';
+    assert_script_run("$quadlet -version");
+
+    # create user container
+    assert_script_run("mkdir -p $systemd_user_path");
+    assert_script_run(qq(printf "$quadlet_container" >"$systemd_user_path/$unit_name"));
+    record_info('Unit', script_output("sudo -su $username $quadlet -user -v -dryrun"));
+
+    # enable user linger
+    assert_script_run("loginctl enable-linger $uid");
+
+    $initial_user_shell = script_output(qq(getent passwd "$user" | cut -d':' -f7));
+    assert_script_run(qq(usermod -s /usr/bin/podmansh "$user"));
+
+    # prepare ssh key for passwordless login
+    my $sshkeypath = "/root/.ssh/id_ed25519";
+    if (script_run(qq(test -e "$sshkeypath.pub")) != 0) {
+        assert_script_run(qq(ssh-keygen -t ed25519 -f "$sshkeypath" -q -N ""));
+    }
+    my $usersshpath = "/home/$user/.ssh";
+    assert_script_run(qq(mkdir -p "$usersshpath" && chown "$user:$user" "$usersshpath"));
+    assert_script_run(qq(tee -a "$usersshpath/authorized_keys" <"$sshkeypath.pub"));
+
+    # pre-pull the image and make sure the service is running
+    script_retry("sudo -su $username sh -c 'cd; podman pull $src_image'", retry => 3, delay => 60, timeout => 180);
+    systemctl("--machine=$username@.host --user start podmansh");
+    systemctl("--machine=$username@.host --user is-active podmansh");
+
+    # change the dafault podmansh non-root shell to match 'text-logged-in-bernhard' needle
+    assert_script_run(qq(sudo -su $user sh -c 'cd; mkdir -p ~/.config/containers/; printf '\\''[podmansh]\\nshell = \"/bin/bash\"\\n'\\'' > ~/.config/containers/containers.conf'));
+
+    # validate podmansh container for multiple login methods
```

Comment:
The tests below look a bit cumbersome to me. They work, but perhaps we could consider another approach:

* Ce could create a canary file in the `Exec` section of the quadlet (e.g. something like `bash -c 'echo "The canary has been planted" > /canary; sleep infinity')
* Grep for the canary file
* Check if the user is root in the container

I think this allows us to achieve the same effect but it might be easier than the suggested commands here.

What do you think?

Reply:
Made some changes, maybe this would do a more reliable verification if running in a container?